precompile/gas_revenue.go

```diff
@@ -0,0 +1,538 @@
+	}
+	unpacked := *abi.ConvertType(res[0], new(common.Address)).(*common.Address)
+	return unpacked, nil
+}
+
+// PackIsRegistered packs [contractAddress] of type common.Address into the appropriate arguments for isRegistered.
+// the packed bytes include selector (first 4 func signature bytes).
+// This function is mostly used for tests.
+func PackIsRegistered(contractAddress common.Address) ([]byte, error) {
+	return GasRevenueABI.Pack("isRegistered", contractAddress)
+}
+
+// PackIsRegisteredOutput attempts to pack given registered of type bool
+// to conform the ABI outputs.
+func PackIsRegisteredOutput(registered bool) ([]byte, error) {
+	return GasRevenueABI.PackOutput("isRegistered", registered)
+}
+
+func isRegistered(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = deductGas(suppliedGas, IsRegisteredGasCost); err != nil {
+		return nil, 0, err
+	}
+	// attempts to unpack [input] into the arguments to the IsRegisteredInput.
+	// Assumes that [input] does not include selector
+	// You can use unpacked [inputStruct] variable in your code
+	inputStruct, err := UnpackIsRegisteredInput(input)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// CUSTOM CODE STARTS HERE
+	isRegistered := IsRegistered(accessibleState.GetStateDB(), inputStruct)
+
+	packedOutput, err := PackIsRegisteredOutput(isRegistered)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Return the packed output and the remaining gas
+	return packedOutput, remainingGas, nil
+}
+
+// PackRegister packs the include selector (first 4 func signature bytes).
+// This function is mostly used for tests.
+func PackRegister() ([]byte, error) {
+	return GasRevenueABI.Pack("register")
+}
+
+// PackRegisterOutput attempts to pack given success of type bool
+// to conform the ABI outputs.
+func PackRegisterOutput(success bool) ([]byte, error) {
+	return GasRevenueABI.PackOutput("register", success)
+}
+
+func register(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = deductGas(suppliedGas, RegisterGasCost); err != nil {
+		return nil, 0, err
+	}
+	if readOnly {
+		return nil, remainingGas, vmerrs.ErrWriteProtection
+	}
+	// no input provided for this function
+
+	// get state of the caller
+	state := accessibleState.GetStateDB().GetState(caller, common.Hash{})
+
+	// if the state is 0, then the caller is not a contract
+	if state.Big().Cmp(big.NewInt(0)) == 0 {
+		return nil, remainingGas, ErrNotSmartContract
+	}
+
+	// check if already registered
+	isRegistered := IsRegistered(accessibleState.GetStateDB(), caller)
+	if isRegistered {
+		return nil, remainingGas, ErrAlreadyRegistered
+	}
+
+	// register
+	SetRegistered(accessibleState.GetStateDB(), caller)
```

Comment:
hm seems like all contracts can register itself to this precompile. If that's the case I don't think you would need a `register` function at all (every contract can be refunded). It seems to me that only permissioned ones should register other contract, shouldn't they? If you plan to do so, you can use `AllowList` to determine who can register target contracts(or all addresses) for this precompile.

Reply:
👆🏻 prev comment